### ISSUE TITLE
Move Contact link/unlink extensions to a dedicated API

### ISSUE THREAD

**Issue by vestrel00** [vestrel00/contacts-android#138]
Currently, linking/unlinking contacts are done via extension functions in `ContactLinks.kt`. 

A sample usage is...

```kotlin
val linkResult = contact1.link(Contacts(context), contact2)
```

## Problem

While this works just fine, it deviates from the API design. Furthermore, it is not possible to add extensions for permissions with this setup. There are currently async extensions...

```kotlin
val linkResult = contact1.linkWithContext(Contacts(context), contact2)
```

The `async` and `permissions` module are mutually exclusive so there is no way to write the following (ugly) function without violating coupling rules,

```kotlin
val linkResult = contact1.linkWithPermissionWithContext(Contacts(context), contact2)
```

## Solution

Contact link/unlink extensions to a dedicated API. Usage would look something like...

```kotlin
val linkResult = Contacts(context)
        .aggregationExceptions()
        .linkWithPermission()
        .contacts(contact1, contact2)
        .commitWithContext()
```

```kotlin
val unlinkResult = Contacts(context)
        .aggregationExceptions()
        .unlinkWithPermission()
        .contact(contact)
        .commitWithContext()
```

Notice `linkWithPermission` and `commitWithContext` can be written in the `permissions` and `async` modules respectively. This is more aligned with the other APIs and is more scalable in case extra parameters need to be added.

- [x] Async extensions
- [x] Permissions extensions

Similar to the approach taken in #119, we'll continue to expose the `link` and `unlink` extension functions provided in `contacts.core.util.ContactLinks.kt` under different function names,

- [x] Rename extensions to `linkDirect` and `unlinkDirect`

Additionally;

- [x] Update sample app
- [x] Update cheatsheet
- [x] Update relevant documentation in code
- [x] Update relevant documentation in docs
- [x] Verify that there are no regressions for related API functions
- [ ] Provide a migration guide as a comment in this issue so that it can be copy-pasted into the release notes

**Comment by vestrel00** [comment]
## Migration guide (to be included in release notes)

[Previously](https://github.com/vestrel00/contacts-android/blob/0.2.4/docs/other/link-unlink-contacts.md), to link /unlink Contacts, you would use one of these extension functions that immediately commits the changes directly into the database. 

```kotlin
contact1.link(contactsApi, contact2, contact3)
```

Now, the above function still exist with a different name.

```kotlin
contact1.linkDirect(contactsApi, contact2, contact3)
```

Additionally, you are now also able to link/unlink using new APIs that match the design of all other existing APIs!

```kotlin
contactsApi
  .aggregationExceptions()
  .linkContacts()
  .contacts(contact1, contact2, contact3)
  .commit()

contactsApi
  .aggregationExceptions()
  .unlinkContact()
  .contact(contact)
  .commit()
```

The advantage of using the new `ContactLink` and `ContactUnlink` APIs is that it allows you to use `[linkContacts|unlinkContact]WithPermission` and `commit[WithContext|Async]` extensions in the same call =)

Read the new documentation for the full guide; https://vestrel00.github.io/contacts-android/other/link-unlink-contacts/

**Comment by vestrel00** [comment]
This is included in [0.3.0](https://github.com/vestrel00/contacts-android/releases/tag/0.3.0)!
